plugins/serializers/msgpack/metric.go

```diff
@@ -0,0 +1,66 @@
+package msgpack
+
+import (
+	"encoding/binary"
+	"time"
+
+	"github.com/tinylib/msgp/msgp"
+)
+
+//go:generate msgp
+
+// Metric is structure to define MessagePack message format
+// will be used by msgp code generator
+type Metric struct {
+	Name   string                 `msg:"name"`
+	Time   MessagePackTime        `msg:"time,extension"`
+	Tags   map[string]string      `msg:"tags"`
+	Fields map[string]interface{} `msg:"fields"`
+}
+
+// MessagePackTime implements the official timestamp extension type
+// https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type
+//
+// tinylib/msgp has been used their own custom extenstion type and the offcial extension
+// is not available. (https://github.com/tinylib/msgp/issues/214)
+type MessagePackTime struct {
+	time time.Time
+}
+
+func init() {
+	msgp.RegisterExtension(-1, func() msgp.Extension { return new(MessagePackTime) })
+}
+
+// ExtensionType implements the Extension interface
+func (*MessagePackTime) ExtensionType() int8 {
+	return -1
+}
+
+// Len implements the Extension interface
+func (*MessagePackTime) Len() int {
+	return 8
+}
+
+// MarshalBinaryTo implements the Extension interface
+func (t *MessagePackTime) MarshalBinaryTo(buf []byte) error {
+	sec := uint64(t.time.Unix())
+	nsec := uint64(t.time.Nanosecond())
+
+	data := nsec<<34 | sec
```

Comment:
How about [UnixNano](https://golang.org/pkg/time/#Time.UnixNano)?

Reply:
The timestamp extension spec uses 34 bits for seconds and 30 bits for nanoseconds so we need the both separately.

And the `UnixNano` is made by adding the both. `(t.unixSec())*1e9 + int64(t.nsec())`

https://golang.org/src/time/time.go#L1137

So, IMO just using `Unix()` and `Nanosecond()` is better for our purpose.

Reply:
Understood. However, I feel better if you mask the second's additional bits before bitwise or... :-)
```suggestion
	sec := uint64(t.time.Unix())
	nsec := uint64(t.time.Nanosecond())

	data := nsec<<34 | (sec & 0x3FFFFFFF)
```

Reply:
How about this? The supported max date is 2514-05-30.

```
	if sec & 0xFFFF_FFFC_0000_0000 != 0 {
		return fmt.Errorf("Time is out of supported range: %v", t)
	}

	data := nsec<<34 | (sec & 0x03_FFFF_FFFF)
```

Reply:
Perfect.